sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java

```diff
@@ -369,4 +380,189 @@ public static <T> T describedToOrigin(DescribedType describedType) {
+        int statusCode = ManagementConstants.UNDEFINED_STATUS_CODE;
+        if (properties.getValue() == null) {
+            return statusCode;
+        }
+
+        Object codeObject = properties.getValue().get(ManagementConstants.STATUS_CODE);
+        if (codeObject == null) {
+            codeObject = properties.getValue().get(ManagementConstants.LEGACY_STATUS_CODE);
+        }
+        if (codeObject != null) {
+            statusCode = (int) codeObject;
+        }
+
+        return statusCode;
+    }
+
+    /**
+     * Get {@link RuleProperties} from {@link DescribedType}.
+     *
+     * @param ruleDescribedType A {@link DescribedType} with rule information.
+     * @return A {@link RuleProperties} contains name, {@link RuleAction} and {@link RuleFilter}.
+     */
+    @SuppressWarnings("unchecked")
+    public static RuleProperties decodeRuleDescribedType(DescribedType ruleDescribedType) {
+        if (ruleDescribedType == null) {
+            return null;
+        }
+        if (!(ruleDescribedType.getDescriptor()).equals(ServiceBusConstants.RULE_DESCRIPTION_NAME)) {
+            return null;
+        }
+
+        RuleProperties ruleProperties = new RuleProperties();
+        if (ruleDescribedType.getDescribed() instanceof ArrayList) {
+            ArrayList<Object> describedRule = (ArrayList<Object>) ruleDescribedType.getDescribed();
+            int count = describedRule.size();
+            if (count-- > 0) {
+                ruleProperties.setFilter(decodeFilter((DescribedType) describedRule.get(0)));
+            }
+
+            if (count-- > 0) {
+                ruleProperties.setAction(decodeRuleAction((DescribedType) describedRule.get(1)));
+            }
+
+            if (count > 0) {
+                ruleProperties.setName((String) describedRule.get(2));
+            }
+        }
+
+        return ruleProperties;
+    }
+
+    /**
+     * Get {@link RuleFilter} from a {@link DescribedType}.
+     *
+     * @param describedFilter A {@link DescribedType} with rule filter information.
+     * @return A {@link RuleFilter}.
+     */
+    @SuppressWarnings("unchecked")
+    private static RuleFilter decodeFilter(DescribedType describedFilter) {
+        if (describedFilter.getDescriptor().equals(ServiceBusConstants.SQL_FILTER_NAME)) {
+            ArrayList<Object> describedSqlFilter = (ArrayList<Object>) describedFilter.getDescribed();
+            if (describedSqlFilter.size() > 0) {
+                return new SqlRuleFilter((String) describedSqlFilter.get(0));
+            }
+        } else if (describedFilter.getDescriptor().equals(ServiceBusConstants.CORRELATION_FILTER_NAME)) {
+            CorrelationRuleFilter correlationFilter = new CorrelationRuleFilter();
+            ArrayList<Object> describedCorrelationFilter = (ArrayList<Object>) describedFilter.getDescribed();
+            int countCorrelationFilter = describedCorrelationFilter.size();
+            if (countCorrelationFilter-- > 0) {
```

Comment:
same here

is this code copied from somewhere?

Reply:
Yeah, these two util functions are copied from Track 1 SDK, I'll do a better pattern and remove hardcode.

Reply:
You can wait for more inputs from Connie or Srikanta before any change, as the coding style is highly opinionated, and they might not agree with me.

Reply:
I agree with weidong on this. It is a bit hard to follow.